### PR TITLE
feat(agents): add market-based agent swarm

### LINKED ISSUE #734: v2.0: Agent Swarms with Emergent Specialization — market-based role negotiation
## Summary
Extend `AgentFederation` so agents *negotiate roles dynamically* via an internal market: each agent bids on incoming tasks based on its historical cost/quality from `CostQualityRouter`. Roles emerge from the market, not from hardcoded orchestration graphs.

## Motivation
Today's multi-agent frameworks (CrewAI, AutoGen, our own `AgentFederation`) require humans to pre-assign roles. This is brittle: as the agent fleet grows, role assignments become stale. A market mechanism self-organizes the team and naturally exploits the strongest specialist for each task.

## Design Sketch

```python
from synapsekit import AgentSwarm, MarketPolicy, BidStrategy

swarm = AgentSwarm(
    agents=[researcher, coder, critic, planner, summarizer, ...],
    market=MarketPolicy(
        bid_strategy=BidStrategy.cost_quality_pareto(),
        auction_type=\"sealed_bid\",
        budget_per_task=10_000,  # tokens
        learning_rate=0.1,  # how fast agents update bids from outcomes
    ),
    registry=AgentRegistry.redis(\"redis://...\"),
)

result = await swarm.execute(\"Write a market analysis on quantum compute startups\")
# Swarm dynamically assembles: planner -> 3 researchers -> critic -> summarizer
# Bids logged. Winners' bid models update from final reward.
```

### Mechanism
1. **Task broadcast** — task posted to swarm with budget
2. **Bid** — each agent submits (estimated_cost, estimated_quality, confidence) using its own `CostQualityRouter` history
3. **Auction** — `MarketPolicy` selects winner(s) — supports single-winner, multi-winner (parallel), and coalition (cooperative) modes
4. **Execute** — winner runs, emits result + actual cost
5. **Settle** — actual reward updates each agent's bid model (online learning)
6. **Reputation** — `AgentRegistry` tracks per-agent track record per task category

## Components
- [ ] `AgentSwarm` orchestrator
- [ ] `MarketPolicy` (auction types: sealed-bid, English, Vickrey)
- [ ] `BidStrategy` interface (cost-quality, Thompson sampling, UCB)
- [ ] `AgentBidder` mixin — adds `.bid(task) -> Bid` to any agent
- [ ] `Reputation` store — Redis-backed, per-(agent, task_category) Bayesian posterior
- [ ] `CoalitionFormer` — agents bid as teams for tasks beyond any single agent
- [ ] Visualization: `swarm.trace_to_mermaid()` — show which agent won which sub-task

## Use Cases
- Heterogeneous fleets (GPT-4o + Claude Opus + local Llama) where each excels at different things
- Cost-bounded research where the market routes to the cheapest sufficient agent
- Emergent specialization: agents that consistently win \"code review\" tasks become the de-facto reviewers without being labeled as such

## Risks
- **Cold start** — new agents have no bid history. Mitigate with optimistic priors + epsilon-greedy exploration.
- **Collusion / runaway bidding** — bound bids to budget; cap concentration (no single agent > 70% of tasks).
- **Cost of the auction itself** — keep bid evaluation O(N), cache reputation lookups

## Acceptance Criteria
- [ ] Demo: 5-agent swarm outperforms hardcoded `AgentFederation` on a 100-task benchmark by ≥15%
- [ ] Cost reduction ≥25% vs. always-route-to-best-model baseline
- [ ] Live dashboard via `PrometheusMetrics`: bids/sec, win rate per agent, avg reward
- [ ] Determinism mode for testing (`market.seed=42`)
- [ ] **Replayable role-allocation receipt** (h/t @clementineCU) — every auction emits an `AgentSwarmReceipt` that can be replayed without trusting the final answer, built in two phases so the reputation prior is never overwritten before it's recorded:
  - **Auction-time fields** (captured before `record_outcome()` runs): `task_id`, candidate agent ids, bid inputs each agent was allowed to see, reputation prior version (the exact snapshot bids were scored against), selected role, rejected roles, budget allocated.
  - **Settlement-time fields** (captured after execution): budget consumed, `outcome_score_source` (`external_override` vs. `self_reported` — collapsed to a two-way grounded/ungrounded split, since agent-returned output fields aren't meaningfully more trustworthy than the bid's own estimate), and `learning_rule_version`.
  - **Grounding is tracked, not silently trusted**: a winning agent's own `estimated_quality`/output-reported quality may appear in the receipt as evidence, but the `ReputationSnapshot` it produces is tagged `grounded=False` unless an externally-supplied reward/quality override was passed to `execute()`. Ungrounded updates still persist (preserves online learning with no external judge wired up) but are discounted/filterable downstream. Hard-refusing ungrounded updates is opt-in via `MarketPolicy.require_grounded_reward=True`, not the default.
  - Spec test: `tests/agents/test_agent_swarm_receipt.py::test_auction_receipt_is_replayable_without_trusting_final_answer` (currently `xfail`; extend with the grounded/ungrounded invariant before removing the marker).

## Related
- Extends `AgentFederation` + `AgentRegistry` (#718)
- Uses `CostQualityRouter` learning signal

### PR DESCRIPTION
**Pull request by DhruvGarg111**
## Summary
- Adds AgentSwarm, MarketPolicy, BidStrategy, auction records, coalition formation, and bid/reputation primitives for market-based agent routing.
- Adds RoutingStrategy.MARKET integration for AgentFederation.run(..., strategy=market) without changing existing routing behavior.
- Adds swarm Prometheus metrics, deterministic market tests, reputation tests, README docs, and a runnable no-API-key example.

## Test Plan
- python -m py_compile src\\synapsekit\\agents\\agent_swarm.py src\\synapsekit\\agents\\agent_registry.py src\\synapsekit\\agents\\federation.py src\\synapsekit\\observability\\metrics.py src\\synapsekit\\agents\\__init__.py src\\synapsekit\\__init__.py examples\\agent_swarm_market.py
- python -m ruff check scoped changed files
- python -m mypy --follow-imports=skip scoped changed source files --no-incremental
- python -m pytest -p pytest_asyncio.plugin -p no:cacheprovider tests\\agents tests\\observability tests\\preflight -q --tb=short (612 passed, 1 skipped)
- python examples\\agent_swarm_market.py

Closes #734

### CODE REVIEW

**Comment by AmitoVrito** [comment]
Merged as part of batch integration PR #763 with production-level test fixes. Thank you @DhruvGarg111!
